typed-th-in-ghc-9/article.md

```diff
@@ -0,0 +1,205 @@
+
+```hs
+>>> :l TH
+
+TH.hs:18:16: error:
+    • Couldn't match type ‘Code m0’ with ‘Q’
+      Expected: Q (TExp LogLevel)
+        Actual: Code m0 LogLevel
+    • In the Template Haskell quotation [|| Production ||]
+      In the expression: [|| Production ||]
+      In a case alternative: Nothing -> [|| Production ||]
+   |
+18 |     Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
+   |                ^^^^^^^^^^^^^^^^^^
+
+TH.hs:20:24: error:
+    • Couldn't match type ‘Code m1’ with ‘Q’
+      Expected: Q (TExp LogLevel)
+        Actual: Code m1 LogLevel
+    • In the Template Haskell quotation [|| Development ||]
+      In the expression: [|| Development ||]
+      In a case alternative: "DEVELOPMENT" -> [|| Development ||]
+   |
+20 |       "DEVELOPMENT" -> [|| Development ||]
+   |                        ^^^^^^^^^^^^^^^^^^^
+
+TH.hs:21:23: error:
+    • Couldn't match type ‘Code m2’ with ‘Q’
+      Expected: Q (TExp LogLevel)
+        Actual: Code m2 LogLevel
+    • In the Template Haskell quotation [|| Production ||]
+      In the expression: [|| Production ||]
+      In a case alternative: "PRODUCTION" -> [|| Production ||]
+   |
+21 |       "PRODUCTION" -> [|| Production ||]
+   |                       ^^^^^^^^^^^^^^^^^^
+Failed, no modules loaded.
+```
+
+In the next sections, we will see how to port this code to GHC 9 without and with backward compatibility with GHC 8.
+
+## Porting old TTH code without backward compatibility
+
+For this section, you need a GHC whose version is at least 9. Make sure to import the appropriate definitions from `template-haskell` as well:
+
+```hs
+import Language.Haskell.TH.Syntax (Code, Q, examineCode, liftCode)
+```
+
+Since we now need to return a value of type `Code`, we change our type signature. And now we can fix the code simply by using two new functions, namely `examineCode` and `liftCode`:
+
+```diff
+-logLevelFromFlag :: Q (TExp LogLevel)
+-logLevelFromFlag = do
++logLevelFromFlag :: Code Q LogLevel
++logLevelFromFlag = liftCode $ do  -- liftCode will turn 'Q' into 'Code'
+  flag <- liftIO $ lookupEnv "LOG_LEVEL"
+  case flag of
++    -- examineCode will turn 'Code' into 'Q'
+-    Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
++    Nothing -> examineCode [|| Production ||]  -- We default to Production if the flag is unset
+    Just level -> case level of
+_      "DEVELOPMENT" -> [|| Development ||]
+_      "PRODUCTION" -> [|| Production ||]
++      "DEVELOPMENT" -> examineCode [|| Development ||]
++      "PRODUCTION" -> examineCode [|| Production ||]
+      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
+```
+
+The errors in the previous section happened because the type returned by the quotations is no longer a `Q`, but `Code` instead. The solution is to use `examineCode` to turn a `Code m a` into a `m (TExp a)`, and then `liftCode` to do the opposite operation: turning a `m (TExp a)` into a `Code m a`.
```

Comment:
Maybe instantiate `m` with `Q` everywhere here, since that’s what you have in your code.

Reply:
Done. I kept the `liftCode` and `examineCode` snippets below unchanged to reflect their actual polymorphic types.